relay-server/src/actors/processor.rs

```diff
@@ -951,77 +962,333 @@ impl EnvelopeProcessorService {
-            self.process_check_ins(state);
-            span::process(state, self.inner.config.clone());
         });
 
         if state.has_event() {
             event::scrub(state)?;
             event::serialize(state)?;
-            if_processing!({
-                span::extract_from_event(state);
-            });
         }
 
         attachment::scrub(state);
+        Ok(())
+    }
+
+    fn process_sessions(&self, state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
+        session::process(state, &self.inner.config);
+        if_processing!(self.inner.config, {
+            self.enforce_quotas(state)?;
+        });
+        Ok(())
+    }
+
+    fn process_security(&self, state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
+        event::extract(state, &self.inner.config)?;
+
+        event::finalize(state, &self.inner.config)?;
+        self.light_normalize_event(state)?;
+        event::filter(state)?;
+
+        if_processing!(self.inner.config, {
+            event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+            self.enforce_quotas(state)?;
+        });
+
+        if state.has_event() {
+            event::scrub(state)?;
+            event::serialize(state)?;
+        }
+        Ok(())
+    }
+
+    fn process_nel(&self, state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
+        event::extract(state, &self.inner.config)?;
+
+        event::finalize(state, &self.inner.config)?;
+        self.light_normalize_event(state)?;
+        event::filter(state)?;
+
+        if_processing!(self.inner.config, {
+            event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+            self.enforce_quotas(state)?;
+        });
+
+        if state.has_event() {
+            event::scrub(state)?;
+            event::serialize(state)?;
+        }
+        Ok(())
+    }
+
+    fn process_unreal(&self, state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
+        if_processing!(self.inner.config, {
+            unreal::expand(state, &self.inner.config)?;
+        });
+
+        event::extract(state, &self.inner.config)?;
+
+        if_processing!(self.inner.config, {
+            unreal::process(state)?;
+            attachment::create_placeholders(state);
+        });
+
+        event::finalize(state, &self.inner.config)?;
+        self.light_normalize_event(state)?;
+        event::filter(state)?;
+
+        if_processing!(self.inner.config, {
+            event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+            self.enforce_quotas(state)?;
+        });
+
+        if state.has_event() {
+            event::scrub(state)?;
+            event::serialize(state)?;
+        }
+        Ok(())
+    }
+
+    fn process_user_reports(
+        &self,
+        state: &mut ProcessEnvelopeState,
+    ) -> Result<(), ProcessingError> {
+        report::process(
```

Comment:
Can we move processing after quotas? 

This is not possible for regular events (errors, transactions, security, nel, ...) since inbound filters must run before quotas and require processed payloads. However, in this case the quotas do not depend on a processed payload, so we can save some work when the project is out of quota.